Introduce fast path in the CPU equal op

houseroad opened the pull request:
Summary: When two tensors share the same storage, and strides, and no other flags, then we should consider this tensors as equal.

Test Plan: buck2 test @//mode/opt //caffe2/test:torch -- --exact 'caffe2/test:torch - test_equal (test_torch.TestTorch)'

Differential Revision: D45282119

ezyang commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -2060,6 +2060,20 @@ bool cpu_equal(const Tensor& self, const Tensor& other) {
+      && self.storage_offset() == other.storage_offset()
+      && self.layout() == other.layout()
+      && self.is_neg() == other.is_neg()
+      && self.is_conj() == other.is_conj()) {
```
You don't need to test these three, they will have been handled before getting here.

houseroad replied:
Just to be safe, I am a bit concerned in some cases, if users directly this cpu_equal function directly, although this should be rare. Keeping these checks shouldn't hurt, or we concern about the overhead for these calls?

ezyang commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -2060,6 +2060,20 @@ bool cpu_equal(const Tensor& self, const Tensor& other) {
+  // TensorIterator, it should be safe to have the following fast path by
+  // ensuring the storage and strides exactly the same.
+  if (self.dtype() == other.dtype()
+      && self.sizes().equals(other.sizes())
```
You don't need to test this, it's tested above

ezyang commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -2060,6 +2060,20 @@ bool cpu_equal(const Tensor& self, const Tensor& other) {
+  // ensuring the storage and strides exactly the same.
+  if (self.dtype() == other.dtype()
+      && self.sizes().equals(other.sizes())
+      && self.strides().equals(other.strides())
```
A fastpath for this would be to instead assert both tensors are contiguous, before checking their strides

ezyang reviewed (approved):
You might also want this to apply to cuda too.

houseroad commented:
Yeah, I will handle the CUDA one in the following PR.

houseroad committed:
Introduce fast path in the CPU equal op (#100024)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/100024

When two tensors share the same storage, and strides, and no other flags, then we should consider this tensors as equal.

We have another approach in https://github.com/pytorch/pytorch/pull/99703, which is directly check equality in the JIT loader. However, we may have to handle the flags like neg/conj explicitly. It's a bit hard to cover all the cases. Per discussion with davidberard98, in the flags like neg/conj should be handled by the dispatcher already (and the TensorIterator logic also proves this), so adding the fast path to CPU and CUDA ops should be a better/safer approach.

Test Plan: buck2 test @//mode/opt //caffe2/test:torch -- --exact 'caffe2/test:torch - test_equal (test_torch.TestTorch)'

Reviewed By: hyuen

Differential Revision: D45282119

fbshipit-source-id: 18e939d236a6d84a79013317db8b2f715f4a3cff
